ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+    COUNT(1) AS request_count,
+
+    COUNT(CASE WHEN (status_code between 200 and 299) THEN 1 ELSE 0 END) as success_count,
+    request_count - success_count AS error_count,
+
+    SUM(response_time_msec) / COUNT(1) AS average_response_time_msec
+  FROM http_request
+  WHERE zone_id = 1 AND minute = date_trunc('minute', now())
+  GROUP BY minute;
+
+We've provided `a data ingest script <http://github.com>`_ you can run to generate example
+data. There are also a few more `example queries <http://github.com>`_ to play around with
+in the github repo.
+
+The above setup will get you pretty far, but has a few drawbacks:
+
+* The dashboard must aggregate every row in the target date range for every query it
+  answers.
+* Storage costs will grow proportionally with the ingest rate and the length of the
+  queryable history.
+
+Rollups
+-------
+
+In order to fix both problems, we have multiple clients who roll up the raw data into a
+pre-aggregated form. Here, we'll aggregat the raw data into other tables which store
+summaries of 1-minute, 1-hour, and 1-day intervals. These might correspond to zoom-levels
+in the dashboard. When the user wants request times for the last month the dashboard can
+read and chart the values for each of the last 30 days, no math required! For the rest of
+this document we'll only talk about the first granularity, the 1-minute one. The github
+repo has `DDL <http://github.com>`_ for the other resolutions.
+
+.. code-block:: sql
+
+  CREATE TABLE http_request_1min (
+        zone_id INT,
+        ingest_time TIMESTAMPTZ, -- which minute this row represents
+
+        error_count INT,
+        success_count INT,
+        request_count INT,
+        average_response_time_msec INT,
+        CHECK (request_count = error_count + success_count)
+  )
+  SELECT master_create_distributed_table('http_requests_1min', 'zone_id', 'hash');
+  SELECT master_create_worker_shards('http_requests_1min', 16, 2);
+  
+  -- indexes aren't automatically created by Citus
+  -- this will create the index on all shards
+  CREATE INDEX ON http_requests_1min (zone_id, ingest_time);
+
+This looks a lot like the previous code block. Most importantly: It also shards on
+``zone_id``, and it also uses 16 shards with 2 replicas of each. Because all three of
+those match, there's a 1-to-1 correspondence between ``http_request`` shards and
+``http_request_1min`` shards, and Citus will place matching shards on the same worker.
+This is called colocation; it makes queries such as joins faster and our rollups possible.
+
+.. image:: /images/colocation.png
+  :alt: colocation in citus
+
+In order to populate ``http_request_1min`` we're going to periodically run the equivalent
+of an INSERT INTO SELECT. Citus doesn't yet support INSERT INTO SELECT on distributed
+tables, so instead we'll run a function on all the workers which runs INSERT INTO SELECT
+on every matching pair of shards (this is possible because the shards are colocated)
```

Comment:
I think it could be useful to add a bit more explanation of the function here, e.g.

This function computes 1-minute aggregates over the raw data in a shard of the http_request table and writes them to the corresponding shard in the http_request_1min table. To ensure that we aggregate the data only once, we keep track of the time range that we've already rolled up in the rollup_threshold table.